lib/xboxkrnl/xboxkrnl.h

```diff
@@ -2386,7 +2386,7 @@ XBAPI VOID NTAPI READ_PORT_BUFFER_UCHAR
     IN ULONG Count
 );
 
-XBAPI volatile OBJECT_TYPE PsThreadObjectType;
+XBAPI OBJECT_TYPE PsThreadObjectType;
```

Comment:
The volatile keyword seems to exist for all kernel exports.
I'm not sure if it's needed (does the compiler recognize `DATA` from an import lib might be modified by other code?).

However, if it's not needed for this `ObjectType`, then it's likely not needed for *any* `ObjectType`.

I feel like this is a premature optimization or incomplete code-cleanup which should be in a separate PR.

Reply:
`PsThreadObjectType` is not expected to change, so the `volatile` is unnecessary either way (we're only passing it's address to kernel APIs anyway). It's removed here because keeping it would mean that code using it (e.g. passing it to `ObReferenceObjectByHandle`, which is done in code submitted here) causes warnings about discarding qualifiers.

Reply:
If we'd get a warning about `volatile` qualifier discarding here, wouldn't we also get one in the drive mounting PR for those object-types in similar contexts?
If we make such assumptions about one `ObjectType`, shouldn't we make it for others, too (for consistency)?

Reply:
The drive mounting PR doesn't touch any `OBJECT_TYPE`s, so no warnings there. Afaik we only use `ObReferenceObjectByHandle` in [GetThreadId](https://github.com/XboxDev/nxdk/blob/ef7735d177c8877b002ef19ac569c9382063978b/lib/winapi/thread.c#L16), but the explicit cast seems to suppress the warning.

Removing `volatile` should be fine for (at least) all `OBJECT_TYPE` exports from the kernel - I didn't want to do more changes than necessary to get this working, but I can integrate the changes for the other occurrences if you want.

Reply:
Oops! I could have sworn that the drive mounting PR used `ObDirectoryObjectType` or `ObSymbolicLinkObjectType`.

I wonder if we should just tolerate the warning for now (= remove this commit), then have a different PR (or Discord discussion) about the volatile keyword.
I think there are good reasons to keep them (things like `KeTickCount`), but also good reasons to remove them (useless if data is passed around / some kernel objects aren't expected to change).
There are also good reasons to remove them everywhere (consistency), but also good reasons to only remove them for some variables (different usage-patterns of different variables).

Removing only *this* volatile keyword seems like a hack that introduces inconsistencies and unexpected user-code (where they manually have to add volatile sometimes; even for similar variables). But removing them everywhere might have unexpected consequences, too.

I'm also not familiar enough with the `volatile` implications to know how effective it really is (does a volatile pointer imply volatile access on the data it points to?).
I'm also not familiar enough with the kernel variables to know which ones *should* be volatile (if any).
I'm also not sure if it makes a difference that these variables are from an import lib (where they are being turned into pointers in the kernel stub anyway).

Reply:
I dropped the commit for now. I'll probably create a separate PR for it sometime later.

> I'm also not sure if it makes a difference that these variables are from an import lib
I'd think so - it seems like this would be the sane thing to do, but it's just a guess. I'll have to read up on the details before even considering changing anything other than the `OBJECT_TYPE`s.

Reply:
>  I'll have to read up on the details before even considering changing anything other than the `OBJECT_TYPE`s.

If you don't create a PR for `volatile` anytime soon, create an issue please (so we don't forget + someone might know how to deal with this).

Reply:
Still needs issue or new PR